plugins/QuickScreenshot/main.lua

```diff
@@ -0,0 +1,115 @@
+--
+--  o The function 'go' is showing a "Save As" dialog before
+--    the user has chosen the region to save. This is because
+--    xournal++ freezes its drawing when os.execute() is called;
+--    in general this is a limitation of os.execute() and it
+--    would be nice to have everything run natively, instead of
+--    using a lua plugin.
+--
+--  o The choice of directory is not saved, and the user must
+--    choose a directory every time. This may or may not be
+--    desirable; it seems to be a GTK+ 3 design choice.
+--
+
+-- Returns 'true' if the system utility exists;
+-- returns 'false' otherwise. This is a dangerous function,
+-- since the argument to popen is not sanitized.
+function existsUtility(utility)
+  if operatingSystem == "Windows" then
+    print("Windows not supported yet.")
+  elseif operatingSystem == "Darwin" then
+    print("macOS not supported yet.")
+  else
+    local f = assert(io.popen("command -v " .. utility))
+    local s = assert(f:read('*a'))
+    f:close()
+    return s ~= ""
+  end
+end
+
+-- Return one of "Windows", "Linux", "Darwin", "FreeBSD",
+-- "OpenBSD" or "NetBSD", etc.
+function findOS()
+  if package.config:sub(1,1) == '\\' then
+     return "Windows"
+  else
+    local f = assert(io.popen("uname -s", 'r'))
+    local s = assert(f:read('*a'))
+    f:close()
+    return s
+  end
+end
+
+-- Register all Toolbar actions and intialize all UI stuff
+function initUi()
+  ref = app.registerUi({["menu"] = "QuickScreenshot", ["callback"] = "go", ["accelerator"] = "<Control>t"});
+  operatingSystem = findOS() -- What's the operating system?
+end
+
+-- This is the callback that gets executed when the user
+-- activates the plugin via the menu or hotkey.
+--
+-- This function calls 'app.saveAs' to create
+-- a "Save As" dialog for the user to choose the
+-- filename where the image should reside. Then
+-- it executes a screenshot utility to capture the
+-- region the user selects.
+--
+function go()
+  local windowsUtilities = {} -- windows programs here
+  local macUtilities = {} -- macOS programs here
+  local elseUtilities =
+    { "gnome-screenshot --area --file="
+    , "scrot --select --silent "
+    }
+
+  -- Launch the "Save As" dialog 
+  local filename = app.saveAs()
+  if not filename then
+    return
+  end
+
+  -- Different operating systems have different
+  -- screenshot utilities, so we check the OS here
+  if operatingSystem == "Windows" then
+    print("Windows not supported yet.")
+  elseif operatingSystem == "Darwin" then
+    print("macOS not supported yet.")
+  else
+    -- This becomes true if at least one screenshot
+    -- utility is available on the system
+    local foundUtility = false
+    -- Check utility availability and use it
+    for i,command in ipairs(elseUtilities) do
+      utilityName = command:match("%S+")
+      if existsUtility(utilityName) then
+        local tmpFilename = os.tmpname()
+        os.execute(command .. tmpFilename .. " &")
```

Comment:
The following code change will make your plugin wait for the user to finish the area selection:

```suggestion
        local file = assert(io.popen(command .. tmpFilename .. " &"))
        file:read('*all')
        file:close()
```
Make this change and launch the "Save as" dialog between this code block and the renaming of the file. That should solve your problem.

Reply:
@rolandlo I tried the suggested code changes, but the issue persists, see the attached image. I think this has to do with the single-thread lua process stalling the xournal++ redrawing. 
![Untitled](https://user-images.githubusercontent.com/7750889/86963476-99bc4700-c132-11ea-9bb9-ec02f058b7cd.png)

Reply:
Hmm, strange, that does not happen on my computer. I have no "pollution" at all, it just works fine. Just to make clear: Have you moved the "Save as dialog" launching right in front of renaming the file? E.g. like this?

```
    for i,command in ipairs(elseUtilities) do
      utilityName = command:match("%S+")
      if existsUtility(utilityName) then
        local tmpFilename = os.tmpname()
        local file = assert(io.popen(command .. tmpFilename .. " &"))
        file:read('*all')
        file:close()
          -- Launch the "Save As" dialog 
        local filename = app.saveAs()
        if not filename then
          return
        end
	      os.rename(tmpFilename, filename)
	      foundUtility = true
	      break
      end
    end

```
More questions:
- Does the pollution happen with both gnome-screenshot and scrot? 
- Can you try taking a screenshot of some area which includes parts outside of the Xournal++ window? So do the screenshot tools pollute parts outside of Xournal++ as well?

Reply:
This does not happen with scrot! The pollution does not happen outside the Xournal++ window. I think scrot does the trick. I will remove gnome-screenshot from the lua code, since that behavior is not acceptable.